Add new nightly dynamic

dhnunes opened the pull request:
This change allows to new nightlies being implemented as needed.

Signed-off-by: Diogo Nunes <dnunes@redhat.com>

netoarmando commented:
@dhnunes if you are not comfortable with the `weekday` parameter, you can use a [var yml file](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#defining-variables-in-files), so you don't need to ask those question every time.

Also, this automation deployment must be idempotent, running it X times will come with the same result every time.

dhnunes commented:
@netoarmando Thanks for the suggestion ! I will answer in line:

> @dhnunes if you are not comfortable with the `weekday` parameter, you can use a [var yml file](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#defining-variables-in-files), so you don't need to ask those question every time.<
 

It's a good idea to use the var file, but providing the different name for each variable ( including time frame and unique identifier to each job ), it can be present in role's default vars.

> Also, this automation deployment must be idempotent, running it X times will come with the same result every time.<

It is, if the unique identifier was not hand-changed, don't matter how many times you ran the playbook only one entry for each line will stand.

f-trivino commented:
I think this PR should include automation for the current nightly flows so if we re-deploy the automation instance the cronjobs are added automatically.

dhnunes commented:
Code changed.

I've added the usage of variables using roles defaults, so, it's consistent across installations and no additional information was requested ( Don't consider the difference in running dialog, this is my devel version of code ):

```
 dnunes@shade $ ansible-playbook ansible/prepare_openclose_pr_tool.yml

Provide an unique identifier for the PR: testing
Branch name to open PR against it: master
Repo owner to create branches [freeipa-pr-ci]: dhnunes
GitHub token: adsasd
Provide the path to a test definition file to use it to open new PRs. For example: ipatests/prci_definition/gating: asdasd
Provide a ssh key that has push access to the git repo: asdasd
Should the PR be open against the upstream repo?. Type no for opening it agaist your own freeipa repo (yes/no): no

PLAY [deploy open_close_pr tool] ***********************************************

TASK [install python and deps for ansible modules] *****************************
changed: [192.168.122.214]

TASK [automation/openclose_pr : install packages dependencies] *****************
[DEPRECATION WARNING]: Invoking "dnf" only once while using a loop via 
squash_actions is deprecated. Instead of using a loop to supply multiple items 
and specifying `name: "{{ item }}"`, please use `name: ['python3-github3py', 
'PyYAML', 'git', 'crontabs']` and remove the loop. This feature will be removed
 in version 2.11. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [192.168.122.214] => (item=['python3-github3py', 'PyYAML', 'git', 'crontabs'])

TASK [automation/openclose_pr : install pip dependencies] **********************
[DEPRECATION WARNING]: Invoking "pip" only once while using a loop via 
squash_actions is deprecated. Instead of using a loop to supply multiple items 
and specifying `name: "{{ item }}"`, please use `name: ['CacheControl', 
'gitpython']` and remove the loop. This feature will be removed in version 
2.11. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [192.168.122.214] => (item=['CacheControl', 'gitpython'])

TASK [automation/openclose_pr : create base directory] *************************
ok: [192.168.122.214]

TASK [automation/openclose_pr : create configuration file] *********************
changed: [192.168.122.214]

TASK [automation/openclose_pr : copy openclose_pr.py script] *******************
ok: [192.168.122.214]

TASK [automation/openclose_pr : cron] ******************************************
changed: [192.168.122.214] => (item={'name': 'testing_master', 'weekdays': '1,3,5', 'hour': 23, 'minute': 0})
changed: [192.168.122.214] => (item={'name': 'testing_f28', 'weekdays': '2,4', 'hour': 23, 'minute': 0})
changed: [192.168.122.214] => (item={'name': 'testing_rawhide', 'weekdays': 6, 'hour': 10, 'minute': 0})

PLAY RECAP *********************************************************************
192.168.122.214            : ok=7    changed=3    unreachable=0    failed=0   

 dnunes@shade $ ansible-playbook ansible/prepare_openclose_pr_tool.yml

Provide an unique identifier for the PR: testing
Branch name to open PR against it: master
Repo owner to create branches [freeipa-pr-ci]: dhnunes
GitHub token: adsasd
Provide the path to a test definition file to use it to open new PRs. For example: ipatests/prci_definition/gating: asdasd
Provide a ssh key that has push access to the git repo: asdasd
Should the PR be open against the upstream repo?. Type no for opening it agaist your own freeipa repo (yes/no): no

PLAY [deploy open_close_pr tool] ***********************************************

TASK [install python and deps for ansible modules] *****************************
changed: [192.168.122.214]

TASK [automation/openclose_pr : install packages dependencies] *****************
[DEPRECATION WARNING]: Invoking "dnf" only once while using a loop via 
squash_actions is deprecated. Instead of using a loop to supply multiple items 
and specifying `name: "{{ item }}"`, please use `name: ['python3-github3py', 
'PyYAML', 'git', 'crontabs']` and remove the loop. This feature will be removed
 in version 2.11. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [192.168.122.214] => (item=['python3-github3py', 'PyYAML', 'git', 'crontabs'])

TASK [automation/openclose_pr : install pip dependencies] **********************
[DEPRECATION WARNING]: Invoking "pip" only once while using a loop via 
squash_actions is deprecated. Instead of using a loop to supply multiple items 
and specifying `name: "{{ item }}"`, please use `name: ['CacheControl', 
'gitpython']` and remove the loop. This feature will be removed in version 
2.11. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [192.168.122.214] => (item=['CacheControl', 'gitpython'])

TASK [automation/openclose_pr : create base directory] *************************
ok: [192.168.122.214]

TASK [automation/openclose_pr : create configuration file] *********************
ok: [192.168.122.214]

TASK [automation/openclose_pr : copy openclose_pr.py script] *******************
ok: [192.168.122.214]

TASK [automation/openclose_pr : cron] ******************************************
ok: [192.168.122.214] => (item={'name': 'testing_master', 'weekdays': '1,3,5', 'hour': 23, 'minute': 0})
ok: [192.168.122.214] => (item={'name': 'testing_f28', 'weekdays': '2,4', 'hour': 23, 'minute': 0})
ok: [192.168.122.214] => (item={'name': 'testing_rawhide', 'weekdays': 6, 'hour': 10, 'minute': 0})

PLAY RECAP *********************************************************************
192.168.122.214            : ok=7    changed=1    unreachable=0    failed=0   
```

netoarmando commented on ansible/roles/automation/openclose_pr/tasks/main.yml:
```diff
@@ -67,18 +67,19 @@
            open_pr
     user: root
     state: present
+  with_items: "{{ nightly_jobs }}"
```
Add a new line

f-trivino replied:
new line here?

netoarmando commented on ansible/roles/automation/openclose_pr/defaults/main.yml:
```diff
@@ -1,2 +1,16 @@
+  - name: testing_rawhide
+    weekdays: 6
+    hour: 10
+    minute: 0
```
Add a new line

dhnunes commented on ansible/prepare_openclose_pr_tool.yml:
```diff
@@ -10,7 +10,7 @@
       prompt: "Provide an unique identifier for the PR"
       private: false
-
+      
```
This appeared because I've force pushed the first commit, in that one, was 3 additional dialogs here, and I need to delete due to review, there is no additional line here, as we can see:

```
  - name: identifier
      prompt: "Provide an unique identifier for the PR"
      private: false
      
    - name: branch
      prompt: "Branch name to open PR against it"
      private: false

    - name: repo_owner
      prompt: "Repo owner to create branches"
      private: false
      default: "freeipa-pr-ci"
```

dhnunes commented:
> > Code changed.
> > I've added the usage of variables using roles defaults, so, it's consistent across installations and no additional information was requested ( Don't consider the difference in running dialog, this is my devel version of code ):
> > ```
> >  dnunes@shade $ ansible-playbook ansible/prepare_openclose_pr_tool.yml
> > 
> > Provide an unique identifier for the PR: testing
> > Branch name to open PR against it: master
> > Repo owner to create branches [freeipa-pr-ci]: dhnunes
> > GitHub token: adsasd
> > Provide the path to a test definition file to use it to open new PRs. For example: ipatests/prci_definition/gating: asdasd
> > Provide a ssh key that has push access to the git repo: asdasd
> > Should the PR be open against the upstream repo?. Type no for opening it agaist your own freeipa repo (yes/no): no
> > 
> > PLAY [deploy open_close_pr tool] ***********************************************
> > 
> > TASK [install python and deps for ansible modules] *****************************
> > changed: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : install packages dependencies] *****************
> > [DEPRECATION WARNING]: Invoking "dnf" only once while using a loop via 
> > squash_actions is deprecated. Instead of using a loop to supply multiple items 
> > and specifying `name: "{{ item }}"`, please use `name: ['python3-github3py', 
> > 'PyYAML', 'git', 'crontabs']` and remove the loop. This feature will be removed
> >  in version 2.11. Deprecation warnings can be disabled by setting 
> > deprecation_warnings=False in ansible.cfg.
> > ok: [192.168.122.214] => (item=['python3-github3py', 'PyYAML', 'git', 'crontabs'])
> > 
> > TASK [automation/openclose_pr : install pip dependencies] **********************
> > [DEPRECATION WARNING]: Invoking "pip" only once while using a loop via 
> > squash_actions is deprecated. Instead of using a loop to supply multiple items 
> > and specifying `name: "{{ item }}"`, please use `name: ['CacheControl', 
> > 'gitpython']` and remove the loop. This feature will be removed in version 
> > 2.11. Deprecation warnings can be disabled by setting 
> > deprecation_warnings=False in ansible.cfg.
> > ok: [192.168.122.214] => (item=['CacheControl', 'gitpython'])
> > 
> > TASK [automation/openclose_pr : create base directory] *************************
> > ok: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : create configuration file] *********************
> > changed: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : copy openclose_pr.py script] *******************
> > ok: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : cron] ******************************************
> > changed: [192.168.122.214] => (item={'name': 'testing_master', 'weekdays': '1,3,5', 'hour': 23, 'minute': 0})
> > changed: [192.168.122.214] => (item={'name': 'testing_f28', 'weekdays': '2,4', 'hour': 23, 'minute': 0})
> > changed: [192.168.122.214] => (item={'name': 'testing_rawhide', 'weekdays': 6, 'hour': 10, 'minute': 0})
> > 
> > PLAY RECAP *********************************************************************
> > 192.168.122.214            : ok=7    changed=3    unreachable=0    failed=0   
> > 
> >  dnunes@shade $ ansible-playbook ansible/prepare_openclose_pr_tool.yml
> > 
> > Provide an unique identifier for the PR: testing
> > Branch name to open PR against it: master
> > Repo owner to create branches [freeipa-pr-ci]: dhnunes
> > GitHub token: adsasd
> > Provide the path to a test definition file to use it to open new PRs. For example: ipatests/prci_definition/gating: asdasd
> > Provide a ssh key that has push access to the git repo: asdasd
> > Should the PR be open against the upstream repo?. Type no for opening it agaist your own freeipa repo (yes/no): no
> > 
> > PLAY [deploy open_close_pr tool] ***********************************************
> > 
> > TASK [install python and deps for ansible modules] *****************************
> > changed: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : install packages dependencies] *****************
> > [DEPRECATION WARNING]: Invoking "dnf" only once while using a loop via 
> > squash_actions is deprecated. Instead of using a loop to supply multiple items 
> > and specifying `name: "{{ item }}"`, please use `name: ['python3-github3py', 
> > 'PyYAML', 'git', 'crontabs']` and remove the loop. This feature will be removed
> >  in version 2.11. Deprecation warnings can be disabled by setting 
> > deprecation_warnings=False in ansible.cfg.
> > ok: [192.168.122.214] => (item=['python3-github3py', 'PyYAML', 'git', 'crontabs'])
> > 
> > TASK [automation/openclose_pr : install pip dependencies] **********************
> > [DEPRECATION WARNING]: Invoking "pip" only once while using a loop via 
> > squash_actions is deprecated. Instead of using a loop to supply multiple items 
> > and specifying `name: "{{ item }}"`, please use `name: ['CacheControl', 
> > 'gitpython']` and remove the loop. This feature will be removed in version 
> > 2.11. Deprecation warnings can be disabled by setting 
> > deprecation_warnings=False in ansible.cfg.
> > ok: [192.168.122.214] => (item=['CacheControl', 'gitpython'])
> > 
> > TASK [automation/openclose_pr : create base directory] *************************
> > ok: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : create configuration file] *********************
> > ok: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : copy openclose_pr.py script] *******************
> > ok: [192.168.122.214]
> > 
> > TASK [automation/openclose_pr : cron] ******************************************
> > ok: [192.168.122.214] => (item={'name': 'testing_master', 'weekdays': '1,3,5', 'hour': 23, 'minute': 0})
> > ok: [192.168.122.214] => (item={'name': 'testing_f28', 'weekdays': '2,4', 'hour': 23, 'minute': 0})
> > ok: [192.168.122.214] => (item={'name': 'testing_rawhide', 'weekdays': 6, 'hour': 10, 'minute': 0})
> > 
> > PLAY RECAP *********************************************************************
> > 192.168.122.214            : ok=7    changed=1    unreachable=0    failed=0   
> > ```
> 
> verify that the cronjobs are added with crontab -l

```
[root@pr-automation ~]# crontab -l
#Ansible: openclose_testing_master
00 23 * * 1,3,5 python3 /root/openclose_pr/open_close_pr.py --config /root/openclose_pr/config.yml --id testing_master --repo_path /root/openclose_pr/freeipa --branch master --prci_config asdasd --pr_against_upstream False open_pr
#Ansible: openclose_testing_f28
00 23 * * 2,4 python3 /root/openclose_pr/open_close_pr.py --config /root/openclose_pr/config.yml --id testing_f28 --repo_path /root/openclose_pr/freeipa --branch master --prci_config asdasd --pr_against_upstream False open_pr
#Ansible: openclose_testing_rawhide
00 10 * * 6 python3 /root/openclose_pr/open_close_pr.py --config /root/openclose_pr/config.yml --id testing_rawhide --repo_path /root/openclose_pr/freeipa --branch master --prci_config asdasd --pr_against_upstream False open_pr
```

f-trivino commented on ansible/roles/automation/openclose_pr/tasks/main.yml:
```diff
@@ -67,18 +67,19 @@
            open_pr
     user: root
     state: present
+  with_items: "{{ nightly_jobs }}"
```
missing spaces?

dhnunes replied:
You mean identation ? with_items should stay at the same level of sub-task

f-trivino replied:
yes, you right! sorry. Let's merge this PR.
